libs/sysaudio/src/jack.zig

```diff
@@ -0,0 +1,274 @@
+                return error.OpeningDevice;
+            dest_ports[i] = dest_name;
+        }
+
+        return .{
+            .jack = .{
+                .allocator = self.allocator,
+                .mutex = .{},
+                .cond = .{},
+                .device = device,
+                .writeFn = writeFn,
+                .client = self.client,
+                .ports = ports,
+                .dest_ports = dest_ports,
+            },
+        };
+    }
+};
+
+pub const Player = struct {
+    allocator: std.mem.Allocator,
+    mutex: std.Thread.Mutex,
+    cond: std.Thread.Condition,
+    device: main.Device,
+    writeFn: main.WriteFn,
+    client: *c.jack_client_t,
+    ports: []const *c.jack_port_t,
+    dest_ports: []const [:0]const u8,
+
+    pub fn deinit(self: *Player) void {
+        self.allocator.free(self.ports);
+        for (self.dest_ports) |d|
+            self.allocator.free(d);
+        self.allocator.free(self.dest_ports);
+    }
+
+    pub fn start(self: *Player) !void {
+        if (c.jack_set_process_callback(self.client, processCallback, self) != 0)
+            return error.CannotPlay;
+
+        if (c.jack_activate(self.client) != 0)
+            return error.CannotPlay;
+
+        for (self.ports) |port, i| {
+            if (c.jack_connect(self.client, c.jack_port_name(port), self.dest_ports[i].ptr) != 0)
+                return error.CannotPlay;
+        }
+    }
+
+    fn processCallback(n_frames: c.jack_nframes_t, self_opaque: ?*anyopaque) callconv(.C) c_int {
+        const self = @ptrCast(*Player, @alignCast(@alignOf(*Player), self_opaque.?));
+        var parent = @fieldParentPtr(main.Player, "data", @ptrCast(*backends.BackendPlayer, self));
+        for (self.channels()) |*ch, i| {
+            ch.*.ptr = @ptrCast([*]u8, c.jack_port_get_buffer(self.ports[i], n_frames));
+        }
+        self.writeFn(parent, n_frames);
+        return 0;
+    }
+
+    pub fn play(self: *Player) !void {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+        for (self.ports) |port, i| {
+            if (c.jack_connect(self.client, c.jack_port_name(port), self.dest_ports[i].ptr) != 0)
+                return error.CannotPlay;
+        }
+    }
+
+    pub fn pause(self: *Player) !void {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+        for (self.ports) |port, i| {
+            if (c.jack_disconnect(self.client, c.jack_port_name(port), self.dest_ports[i].ptr) != 0)
+                return error.CannotPause;
+        }
+    }
+
+    pub fn paused(self: *Player) bool {
+        self.mutex.lock();
+        defer self.mutex.unlock();
+        for (self.ports) |port, i| {
+            if (c.jack_port_connected_to(port, self.dest_ports[i].ptr) == 1)
+                return false;
+        }
+        return true;
+    }
+
+    pub fn setVolume(self: *Player, vol: f32) !void {
+        _ = self;
+        _ = vol;
+        @panic("incompatible backend");
```

Comment:
Is panic desirable behavior here? Seems like this should be a documented no-op instead maybe.